packages/core/src/services/providerLogs/addMessages.ts

```diff
@@ -101,85 +97,50 @@ async function streamMessageResponse({
   documentLogUuid?: string
 }) {
   try {
+    await checkFreeProviderQuota({
+      workspace,
+      provider,
+    }).then((r) => r.unwrap())
+
     const providerProcessor = new ProviderProcessor({
       source,
       documentLogUuid,
       config,
       apiProvider: provider,
       messages,
+      saveSyncProviderLogs: true,
     })
+    const stepStartTime = Date.now()
     const result = await ai({
-      workspace,
       messages,
       config,
       provider,
-    })
-
-    for await (const value of streamToGenerator<
-      TextStreamPart<Record<string, CoreTool>> | ObjectStreamPart<unknown>
-    >(result.data.fullStream)) {
-      enqueueChainEvent(controller, {
-        event: StreamEventTypes.Provider,
-        data: value as unknown as ProviderData,
-      })
-    }
-
+    }).then((r) => r.unwrap())
+    const streamConsumedResult = await consumeStream({ controller, result })
     const response = await providerProcessor.call({
       aiResult: result,
-      saveSyncProviderLogs: true,
+      startTime: stepStartTime,
+      streamConsumedResult,
     })
     const providerLog = response.providerLog!
-    const isStreamText = response.streamType === 'text'
-    const isStreamObject = response.streamType === 'object'
-    const textContent = isStreamText
-      ? response.text
-      : isStreamObject
-        ? objectToString(response.object)
-        : ''
-    enqueueChainEvent(controller, {
-      event: StreamEventTypes.Latitude,
-      data: {
-        type: ChainEventTypes.Complete,
-        config: { ...config, provider: provider.name, model: config.model },
-        documentLogUuid,
-        messages: [
-          {
-            role: MessageRole.assistant,
-            toolCalls: isStreamText ? response.toolCalls : [],
-            content: textContent,
-          },
-        ],
-        response: {
-          ...response,
-          providerLog,
-          text: textContent,
-        },
+    const text = parseResponseText(response)
+    ChainStreamConsumer.chainCompleted({
+      controller,
+      response: { ...response, providerLog, text },
+      config: {
+        ...config,
+        provider: provider.name,
+        model: config.model,
       },
     })
 
-    controller.close()
-
-    return {
-      ...response,
-      providerLog,
-      text: textContent,
-    }
+    return { ...response, providerLog, text }
   } catch (e) {
-    const error = e as Error
-    enqueueChainEvent(controller, {
-      event: StreamEventTypes.Latitude,
-      data: {
-        type: ChainEventTypes.Error,
-        error: {
-          name: error.name,
-          message: error.message,
-          stack: error.stack,
-        },
-      },
+    const error = ChainStreamConsumer.chainError({
```

Comment:
this enqueues the chain error and closes the controller no?

Reply:
Yes